js/reference.js

```diff
@@ -2117,8 +2136,13 @@
                     for (var j=0; j<keyColumns.length; j++) {
                         var keyCol = keyColumns[j],
                         data = tupleData[mapping.get(keyCol).name];
-
-                        if (data === undefined || data === null) return null; // tuple data for keyCol doesn't exist
+                        if (data === undefined || data === null) {
+                            // this returns an error, if a value was shown in the UI with a null value for the right side of the
+                            //         uniqueness constraint, something about the model is broken and further UX actions should not continue
+                            // NOTE: include more context in the error message to communicate information about the parent, assocation, and leaf tables
+                            var err = new module.InvalidInputError("One or more " + associationRef.displayname.value + " tuples have a null value for " + mapping.get(keyCol).name);
+                            defer.reject(err);
```

Comment:
Return the promise right after rejecting. Based on the behavior that I saw in the past, just rejecting will not terminate the code path. You could also write them in the same line:

```
return defer.reject(err), defer.promise;
```